Safer grip and add ungrip

ykyohei opened the pull request:
## Description
- Add some documentations
- Make grip safer
- Add ungrip

## Motivation and Context
https://github.com/simonsobs/chwp-discussions/discussions/17

## How Has This Been Tested?
We need to test

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

bbixler500 commented on socs/agents/hwp_gripper/agent.py:
```diff
@@ -460,6 +515,10 @@ def run_and_append(func, *args, **kwargs):
 
+        if (not all(finished)) or any(aborted):
+            self.log.error('HWP is not correctly gripped')
+            run_and_append(self.client.home, job='grip', check_shutdown=check_shutdown)
```
For this to work consistently, you should probably turn on the `force` function before moving home

BrianJKoopman commented on socs/agents/hwp_gripper/agent.py:
```diff
@@ -315,7 +316,9 @@ def act(self, session, params=None):
         """is_cold(value=False)
 
-        **Task** - Set the code to operate in warm/cold grip configuration
+        **Task** - Set the limit switches to warm/cold grip configuration
+        warm grip configuration: enable both warm and cold limit switches.
+        cold grip configuration: enable only cold limit switches.
```
The formatting of this is a bit messy when rendered in sphinx. "Set the limit switches to warm/cold grip configuration" runs right into "warm grip configuration" to read "grip configuration warm grip configuration". At minimum add a period to the end of the first line. I would place the meaning of "warm grip" and "cold grip" in either the "Parameters" section where `value` is described, or down in the "Notes" section where the description of querying "the correct set of limit switches" is.

BrianJKoopman commented on socs/agents/hwp_gripper/agent.py:
```diff
@@ -470,6 +525,87 @@ def run_and_append(func, *args, **kwargs):
+            Session object for current operation. This function will add to
+            session data.
+
+        Notes:
+            The most recent data collected is stored in session data in the
+            structure:
+
+                >>> response.session['responses']
+                [{'result': True, 'log': [.., .., etc]},
+                                    ..
+                 {'result': True, 'log': [.., .., etc]}]
```
Thanks for including this `session.data` documentation. It should be moved up to the `ungrip()` task, otherwise it will not be visible on readthedocs.

BrianJKoopman commented on socs/agents/hwp_gripper/agent.py:
```diff
@@ -470,6 +525,87 @@ def run_and_append(func, *args, **kwargs):
+        grippers as much as possible.
+        """
+
+        self._ungrip_hwp(session, check_shutdown=True)
```
Rather than passing the `session` object to this helper function, it'd be better (and fairly straightforward) to set the `session.data` within the `ungrip()` function after returning `data` from `_ungrip_hwp(check_shutdown=True)`. So it would look like:
```
session.data = self._ungrip_hwp(check_shutdown=True)
```

This decouples `_ungrip_hwp()` a bit more, and you're already returning the `data` from that function.

BrianJKoopman commented on socs/agents/hwp_gripper/agent.py:
```diff
@@ -470,6 +525,87 @@ def run_and_append(func, *args, **kwargs):
+
+        # check limit switch state
+        self.monitor_state(session)
+        limit_switch_state = (not session['data']['state']['act1_limit_warm_grip_state']) & \
+                             (not session['data']['state']['act2_limit_warm_grip_state']) & \
+                             (not session['data']['state']['act3_limit_warm_grip_state'])
+
+        # We should stop schedule if we have an error in this task
+        assert limit_switch_state, "Failed to ungrip HWP."
```
I'd like to see this check moved up into `ungrip()`, and rather than raising an `AssertionError` for force a task failure, do a `return False, "Failed to ungrip HWP"`, which cleanly indicates Task failure.

BrianJKoopman commented on socs/agents/hwp_gripper/agent.py:
```diff
@@ -659,6 +795,8 @@ def make_parser(parser=None):
                         help='IP of Gripper Beaglebone')
     pgroup.add_argument('--control-port', type=int, default=8041,
                         help='Port for actuator control as set by the Beaglebone code')
+    pgroup.add_argument('--warm-grip-distance', nargs='+', default=[10.0, 10.0, 10.0],
+                        help='Nominal distance for warm grip position')
```
If this'll be commonly used in SCFs, can you add it to the example in `docs/agents/hwp_gripper.rst`?

ykyohei replied:
This is solved in latest ocs thanks to Jack!

BrianJKoopman reviewed (changes requested):
A few comments from me related to documentation and the handling of session data in the new ungrip code. I'm also wondering if this has been tested yet.

ykyohei commented:
I finished incorporating the comments from @BrianJKoopman  and testing `grip` and `ungrip` in satp3.
I will ask @bbixler500  to review again because I changed the grip function little bit.

- Move warm-grip-distance from -5 mm in the first step
- Move in by 0.1 mm step, until warm-limit switch trigger, warm-grip-distance +1 mm at maximum
- Added adjustment process. I would like to implement this process because satp3 limit switches are known to be slightly mis-aligned from ideal location little bit. You can do `adjustment-distance = [0, 0, 0]` if you do not adjustment.

bbixler500 reviewed (approved):
Yep looks good to me

BrianJKoopman reviewed (approved):
Changes related to my comments look good. @jlashner, did you still want to take a look at this?

jlashner reviewed (approved):
Hi Kyohei, I read this through and think it looks great, so I approve to merge as long as its been tested!
